crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,530 @@
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use crate::operation_text::hash;
+use common_lang_types::{
+    ArtifactFileName, ArtifactHash, ArtifactPathAndContent, FileContent, FileSystemOperation,
+    SelectableName, ServerObjectEntityName,
+};
+use isograph_config::PersistedDocumentsHashAlgorithm;
+
+#[derive(Debug, Clone, Default)]
+pub struct FileSystemState {
+    root_files: HashMap<ArtifactFileName, (FileContent, ArtifactHash)>,
+    nested_files: HashMap<
+        ServerObjectEntityName,
+        HashMap<SelectableName, HashMap<ArtifactFileName, (FileContent, ArtifactHash)>>,
+    >,
+}
+
+impl FileSystemState {
+    pub fn insert(&mut self, path_and_content: &ArtifactPathAndContent) {
+        let value = (
+            FileContent::from(path_and_content.file_content.clone()),
+            ArtifactHash::from(hash(
+                &path_and_content.file_content,
+                PersistedDocumentsHashAlgorithm::Md5,
+            )),
+        );
+
+        match &path_and_content.artifact_path.type_and_field {
+            Some(type_and_field) => {
+                self.nested_files
+                    .entry(type_and_field.parent_object_entity_name)
+                    .or_default()
+                    .entry(type_and_field.selectable_name)
+                    .or_default()
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+            None => {
+                self.root_files
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+        }
+    }
+
+    pub fn from_artifacts(artifacts: Vec<ArtifactPathAndContent>) -> Self {
+        let mut state = Self::default();
+        for artifact in artifacts {
+            state.insert(&artifact);
+        }
+        state
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.root_files.is_empty() && self.nested_files.is_empty()
+    }
+
+    // Computes filesystem operations needed to transform this state into the target state.
+    // Returns operations to create, update, and delete files and directories. Files are updated
+    // only when their hash differs. If the current state is empty, emits a DeleteDirectory
+    // operation for the artifact root followed by recreation of all files. Empty directories
+    // are automatically removed after their files are deleted.
+    pub fn diff(&self, new: &Self, artifact_directory: &PathBuf) -> Vec<FileSystemOperation> {
+        let mut operations: Vec<FileSystemOperation> = Vec::new();
+
+        if self.is_empty() {
+            operations.push(FileSystemOperation::DeleteDirectory(
+                artifact_directory.to_path_buf(),
+            ));
+
+            for (server_object, selectables) in &new.nested_files {
+                let server_object_path = artifact_directory.join(server_object.to_string());
+                operations.push(FileSystemOperation::CreateDirectory(
```

Comment:
I think we should skip this `CreateDirectory`, as `CreateDirectory` already creates calls `create_dir_all`, which will:
/// Recursively create a directory and all of its parent components if they
/// are missing.